File: JavaScript/examples/string_utils_example.js
/**
 * String Utils Example
 * 
 * Demonstrates usage of the StringUtils module.
 * 
 * Run with: node string_utils_example.js
 */

const StringUtils = require('../string_utils/mod.js');

console.log('========================================');
console.log('StringUtils Examples');
console.log('========================================\n');

// Empty/Blank Checks
console.log('--- Empty/Blank Checks ---');
console.log('isBlank("hello"):', StringUtils.isBlank('hello'));
console.log('isBlank("  "):', StringUtils.isBlank('  '));
console.log('isBlank(null):', StringUtils.isBlank(null));
console.log('isNotBlank("hello"):', StringUtils.isNotBlank('hello'));
console.log();

// Trimming and Whitespace
console.log('--- Trimming and Whitespace ---');
console.log('trim("  hello  "):', StringUtils.trim('  hello  '));
console.log('trimLeft("  hello  "):', StringUtils.trimLeft('  hello  '));
console.log('trimRight("  hello  "):', StringUtils.trimRight('  hello  '));
console.log('removeWhitespace("h e l l o"):', StringUtils.removeWhitespace('h e l l o'));
console.log('normalizeWhitespace("hello   world"):', StringUtils.normalizeWhitespace('hello   world'));
console.log();

// Case Conversion
console.log('--- Case Conversion ---');
console.log('toLowerCase("HELLO"):', StringUtils.toLowerCase('HELLO'));
console.log('toUpperCase("hello"):', StringUtils.toUpperCase('hello'));
console.log('capitalize("hello"):', StringUtils.capitalize('hello'));
console.log('uncapitalize("Hello"):', StringUtils.uncapitalize('Hello'));
console.log('toTitleCase("hello world"):', StringUtils.toTitleCase('hello world'));
console.log('swapCase("Hello"):', StringUtils.swapCase('Hello'));
console.log();

// Substring Operations
console.log('--- Substring Operations ---');
console.log('truncate("hello world", 8):', StringUtils.truncate('hello world', 8));
console.log('truncate("hello world", 8, " >>"):', StringUtils.truncate('hello world', 8, ' >>'));
console.log('substringBetween("hello [world] there", "[", "]"):', StringUtils.substringBetween('hello [world] there', '[', ']'));
console.log('substringAfter("hello-world", "-"):', StringUtils.substringAfter('hello-world', '-'));
console.log('substringBefore("hello-world", "-"):', StringUtils.substringBefore('hello-world', '-'));
console.log('substringAfterLast("a-b-c", "-"):', StringUtils.substringAfterLast('a-b-c', '-'));
console.log('substringBeforeLast("a-b-c", "-"):', StringUtils.substringBeforeLast('a-b-c', '-'));
console.log();

// Prefix/Suffix Operations
console.log('--- Prefix/Suffix Operations ---');
console.log('startsWith("hello world", "hello"):', StringUtils.startsWith('hello world', 'hello'));
console.log('startsWith("Hello world", "hello", true):', StringUtils.startsWith('Hello world', 'hello', true));
console.log('endsWith("hello world", "world"):', StringUtils.endsWith('hello world', 'world'));
console.log('removePrefix("hello world", "hello "):', StringUtils.removePrefix('hello world', 'hello '));
console.log('removeSuffix("hello world", " world"):', StringUtils.removeSuffix('hello world', ' world'));
console.log();

// Counting and Searching
console.log('--- Counting and Searching ---');
console.log('countMatches("hello hello", "hello"):', StringUtils.countMatches('hello hello', 'hello'));
console.log('contains("hello world", "world"):', StringUtils.contains('hello world', 'world'));
console.log('contains("Hello World", "world", true):', StringUtils.contains('Hello World', 'world', true));
console.log('indexOf("hello world", "world"):', StringUtils.indexOf('hello world', 'world'));
console.log('lastIndexOf("hello hello", "hello"):', StringUtils.lastIndexOf('hello hello', 'hello'));
console.log();

// Replacement
console.log('--- Replacement ---');
console.log('replaceAll("hello hello", "hello", "hi"):', StringUtils.replaceAll('hello hello', 'hello', 'hi'));
console.log('replaceFirst("hello hello", "hello", "hi"):', StringUtils.replaceFirst('hello hello', 'hello', 'hi'));
console.log('replaceLast("hello hello", "hello", "hi"):', StringUtils.replaceLast('hello hello', 'hello', 'hi'));
console.log();

// Padding
console.log('--- Padding ---');
console.log('padLeft("5", 3, "0"):', StringUtils.padLeft('5', 3, '0'));
console.log('padRight("5", 3, "0"):', StringUtils.padRight('5', 3, '0'));
console.log('center("hi", 6, "-"):', StringUtils.center('hi', 6, '-'));
console.log();

// Reversal and Repetition
console.log('--- Reversal and Repetition ---');
console.log('reverse("hello"):', StringUtils.reverse('hello'));
console.log('repeat("ab", 3):', StringUtils.repeat('ab', 3));
console.log();

// Splitting and Joining
console.log('--- Splitting and Joining ---');
console.log('split("a,b,c", ","):', JSON.stringify(StringUtils.split('a,b,c', ',')));
console.log('lines("line1\\nline2\\nline3"):', JSON.stringify(StringUtils.lines('line1\nline2\nline3')));
console.log('join(["a", "b", "c"], "-"):', StringUtils.join(['a', 'b', 'c'], '-'));
console.log();

// Validation
console.log('--- Validation ---');
console.log('isValidEmail("test@example.com"):', StringUtils.isValidEmail('test@example.com'));
console.log('isValidEmail("invalid"):', StringUtils.isValidEmail('invalid'));
console.log('isValidUrl("https://example.com"):', StringUtils.isValidUrl('https://example.com'));
console.log('isValidIPv4("192.168.1.1"):', StringUtils.isValidIPv4('192.168.1.1'));
console.log('isNumeric("123.45"):', StringUtils.isNumeric('123.45'));
console.log('isInteger("123"):', StringUtils.isInteger('123'));
console.log('isAlpha("abc"):', StringUtils.isAlpha('abc'));
console.log('isAlphanumeric("abc123"):', StringUtils.isAlphanumeric('abc123'));
console.log();

// Naming Conventions
console.log('--- Naming Conventions ---');
console.log('toCamelCase("hello-world"):', StringUtils.toCamelCase('hello-world'));
console.log('toCamelCase("hello_world"):', StringUtils.toCamelCase('hello_world'));
console.log('toPascalCase("hello-world"):', StringUtils.toPascalCase('hello-world'));
console.log('toSnakeCase("helloWorld"):', StringUtils.toSnakeCase('helloWorld'));
console.log('toKebabCase("helloWorld"):', StringUtils.toKebabCase('helloWorld'));
console.log();

// Random Generation
console.log('--- Random Generation ---');
console.log('random(10):', StringUtils.random(10));
console.log('randomAlphanumeric(10):', StringUtils.randomAlphanumeric(10));
console.log('randomNumeric(10):', StringUtils.randomNumeric(10));
console.log('randomAlphabetic(10):', StringUtils.randomAlphabetic(10));
console.log('randomPassword(16):', StringUtils.randomPassword(16));
console.log();

// URL Encoding
console.log('--- URL Encoding ---');
console.log('urlEncode("hello world"):', StringUtils.urlEncode('hello world'));
console.log('urlDecode("hello%20world"):', StringUtils.urlDecode('hello%20world'));
console.log();

// Slug Generation
console.log('--- Slug Generation ---');
console.log('slugify("Hello World!"):', StringUtils.slugify('Hello World!'));
console.log('slugify("My Blog Post Title"):', StringUtils.slugify('My Blog Post Title'));
console.log();

// HTML Operations
console.log('--- HTML Operations ---');
console.log('stripHtml("<p>hello</p>"):', StringUtils.stripHtml('<p>hello</p>'));
console.log('escapeHtml("<script>alert(1)</script>"):', StringUtils.escapeHtml('<script>alert(1)</script>'));
console.log('unescapeHtml("&lt;script&gt;"):', StringUtils.unescapeHtml('&lt;script&gt;'));
console.log();

// Comparison
console.log('--- Comparison ---');
console.log('equals("hello", "hello"):', StringUtils.equals('hello', 'hello'));
console.log('equals("Hello", "hello", true):', StringUtils.equals('Hello', 'hello', true));
console.log('compare("a", "b"):', StringUtils.compare('a', 'b'));
console.log();

// Default Values
console.log('--- Default Values ---');
console.log('defaultString("", "default"):', StringUtils.defaultString('', 'default'));
console.log('defaultString("hello", "default"):', StringUtils.defaultString('hello', 'default'));
console.log('defaultIfEmpty("", "default"):', StringUtils.defaultIfEmpty('', 'default'));
console.log('defaultIfEmpty("  ", "default"):', StringUtils.defaultIfEmpty('  ', 'default'));
console.log();

console.log('========================================');
console.log('Examples completed!');
console.log('========================================');

File: Fortran/examples/string_utils_example.f90
!! AllToolkit - Fortran String Utilities Example
!! Demonstrates usage of string_utils module

program string_utils_example
    use string_utils
    implicit none
    
    character(len=100) :: str, result_str
    integer :: count_val, pos
    
    print *, '========================================'
    print *, 'Fortran String Utilities Examples'
    print *, '========================================'
    print *, ''
    
    ! Example 1: Empty/Blank Checks
    print *, '--- Example 1: Empty/Blank Checks ---'
    str = '  '
    print *, 'String: "', trim(str), '"'
    print *, 'is_blank: ', is_blank(str)
    print *, 'is_empty: ', is_empty(str)
    str = 'hello'
    print *, 'String: "', trim(str), '"'
    print *, 'is_not_blank: ', is_not_blank(str)
    print *, ''
    
    ! Example 2: Trimming and Whitespace
    print *, '--- Example 2: Trimming and Whitespace ---'
    str = '  hello world  '
    print *, 'Original: "', trim(str), '"'
    print *, 'str_trim: "', trim(str_trim(str)), '"'
    str = 'h e l l o'
    print *, 'Original: "', trim(str), '"'
    print *, 'remove_whitespace: "', trim(remove_whitespace(str)), '"'
    str = 'hello   world'
    print *, 'Original: "', trim(str), '"'
    print *, 'normalize_whitespace: "', trim(normalize_whitespace(str)), '"'
    print *, ''
    
    ! Example 3: Case Conversion
    print *, '--- Example 3: Case Conversion ---'
    str = 'Hello World'
    print *, 'Original: "', trim(str), '"'
    print *, 'to_lower: "', trim(to_lower(str)), '"'
    print *, 'to_upper: "', trim(to_upper(str)), '"'
    print *, 'capitalize: "', trim(capitalize('hello')), '"'
    print *, 'swap_case: "', trim(swap_case(str)), '"'
    print *, ''
    
    ! Example 4: Substring Operations
    print *, '--- Example 4: Substring Operations ---'
    str = 'hello world'
    print *, 'String: "', trim(str), '"'
    print *, 'substring_after "hello ": "', trim(substring_after(str, 'hello ')), '"'
    print *, 'substring_before " world": "', trim(substring_before(str, ' world')), '"'
    str = 'hello hello world'
    print *, 'String: "', trim(str), '"'
    print *, 'substring_after_last "hello ": "', trim(substring_after_last(str, 'hello ')), '"'
    str = '<div>content</div>'
    print *, 'String: "', trim(str), '"'
    print *, 'substring_between: "', trim(substring_between(str, '<div>', '</div>')), '"'
    str = 'hello world'
    print *, 'truncate to 8 with "...": "', trim(truncate(str, 8, '...')), '"'
    print *, ''
    
    ! Example 5: Prefix/Suffix Operations
    print *, '--- Example 5: Prefix/Suffix Operations ---'
    str = 'hello world'
    print *, 'String: "', trim(str), '"'
    print *, 'starts_with "hello": ', starts_with(str, 'hello')
    print *, 'ends_with "world": ', ends_with(str, 'world')
    print *, 'starts_with_ignore_case "HELLO": ', starts_with_ignore_case(str, 'HELLO')
    print *, 'remove_prefix "hello ": "', trim(remove_prefix(str, 'hello ')), '"'
    print *, 'remove_suffix " world": "', trim(remove_suffix(str, ' world')), '"'
    print *, ''
    
    ! Example 6: Search and Count
    print *, '--- Example 6: Search and Count ---'
    str = 'hello hello world'
    print *, 'String: "', trim(str), '"'
    count_val = count_substring(str, 'hello')
    print *, 'count_substring "hello": ', count_val
    print *, 'contains_substring "world": ', contains_substring(str, 'world')
    pos = index_of(str, 'world')
    print *, 'index_of "world": ', pos
    pos = last_index_of(str, 'hello')
    print *, 'last_index_of "hello": ', pos
    print *, ''
    
    ! Example 7: Replacement
    print *, '--- Example 7: Replacement ---'
    str = 'hello world'
    print *, 'Original: "', trim(str), '"'
    print *, 'replace_first "world" -> "universe": "', &
             trim(replace_first(str, 'world', 'universe')), '"'
    str = 'hello hello hello'
    print *, 'Original: "', trim(str), '"'
    print *, 'replace_all "hello" -> "hi": "', &
             trim(replace_all(str, 'hello', 'hi')), '"'
    print *, ''
    
    ! Example 8: Padding
    print *, '--- Example 8: Padding ---'
    str = '42'
    print *, 'Original: "', trim(str), '"'
    print *, 'pad_left to 5 with "0": "', trim(pad_left(str, 5, '0')), '"'
    print *, 'pad_right to 5 with "0": "', trim(pad_right(str, 5, '0')), '"'
    print *, 'center in 6 with spaces: "', trim(center('hi', 6, ' ')), '"'
    print *, ''
    
    ! Example 9: Reversal and Repetition
    print *, '--- Example 9: Reversal and Repetition ---'
    str = 'hello'
    print *, 'Original: "', trim(str), '"'
    print *, 'reverse_string: "', trim(reverse_string(str)), '"'
    print *, 'repeat_string 3 times: "', trim(repeat_string('hi', 3)), '"'
    print *, ''
    
    ! Example 10: Validation
    print *, '--- Example 10: Validation ---'
    print *, 'is_numeric "123": ', is_numeric('123')
    print *, 'is_numeric "3.14": ', is_numeric('3.14')
    print *, 'is_numeric "abc": ', is_numeric('abc')
    print *, 'is_integer_str "123": ', is_integer_str('123')
    print *, 'is_alpha "abc": ', is_alpha('abc')
    print *, 'is_alphanumeric "abc123": ', is_alphanumeric('abc123')
    print *, 'is_valid_email "test@example.com": ', is_valid_email('test@example.com')
    print *, 'is_valid_email "invalid": ', is_valid_email('invalid')
    print *, ''
    
    ! Example 11: Utility Functions
    print *, '--- Example 11: Utility Functions ---'
    str = ''
    print *, 'default_if_blank for empty: "', &
             trim(default_if_blank(str, 'default')), '"'
    str = 'hello'
    print *, 'default_if_blank for "hello": "', &
             trim(default_if_blank(str, 'default')), '"'
    print *, 'str_equals "hello" == "hello": ', str_equals('hello', 'hello')
    print *, 'str_equals_ignore_case "Hello" == "HELLO": ', &
             str_equals_ignore_case('Hello', 'HELLO')
    print *, ''
    
    print *, '========================================'
    print *, 'Examples completed!'
    print *, '========================================'

end program string_utils_example
